### PR TITLE
Refactor build Config. 

### PR DESCRIPTION
**Pull request by cedricgautier**

### CODE REVIEW

File: back/config/config.go [hunk 1]
```diff
@@ -13,115 +13,52 @@ import (
...
+		return Config{}, errors.Errorf("Couldn't load .env file %s", err)
 	}
 
 	config.Handler, err = loadClientHandler(config)
 	if err != nil {
 		return Config{}, errors.Wrapf(err, "InitializeConfig, unable to load client handler")
 	}
 
-	config, err = buildEnvironmentConfig(config)
+	config, err = selectConfigProcessor[config.Env.BuildProduction](config)
```
Comment (avetisk) [inline back/config/config.go]: You're shadowing your `config` param from line 16.

File: back/config/config.go [hunk 2]
```diff
@@ -13,115 +13,52 @@ import (
...
-			return Config{}, errors.Wrapf(err, "buildEnvironmentConfig, unable to initialize production database")
-		}
+	config.Processor = configProcessor
 
-		log.Info("Production Environment has been loaded")
+	if err = initDatabase(config, env); err != nil {
+		return Config{}, errors.Wrapf(err, "buildEnvironmentConfig, unable to initialize database")
 	}
 
+	log.Info("Which Environment has been loaded")
```
Comment (avetisk) [inline back/config/config.go]: > "Which Environment has been loaded"

That's a good question, which one? 😅

**Review by avetisk — COMMENTED** [review]
Much better! 

Try to factor your error handlers too. Lots of messages are quite similar.

**Review by avetisk — APPROVED** [review]

File: back/config/config.go (hunk 2 above)
Comment (cedricgautier) [reply back/config/config.go]: You're right, let me change that!

Comment (cedricgautier) [reply back/config/config.go]: Change [is here](https://github.com/KrispyTech/airneis/pull/51/commits/c65c96d9b1e15f9608a724e67393e043183d2dbc#diff-c4a1534e01e91809b6c81511a81473dd6b6925fb4a882c91274909e82cc87fd3R61 )

File: back/config/config.go (hunk 1 above)
Comment (cedricgautier) [reply back/config/config.go]: [Changes made here](https://github.com/KrispyTech/airneis/pull/51/commits/32328d37db50e8e3a951b083e7094022842e7e9f)

File: back/config/config.go [hunk 3]
```diff
@@ -13,115 +14,53 @@ import (
...
+		return Config{}, errors.Errorf("loadConfig, unable to read config file %s", err)
+	}
 
-	} else {
-		config, err := loadConfig(config, productionEnv)
-		if err != nil {
-			return Config{}, errors.Wrapf(err, "buildEnvironmentConfig, unable to load config")
-		}
+	if err = yaml.Unmarshal(configFile, &configProcessor); err != nil {
+		return Config{}, errors.Errorf("loadConfig, unable to unmarshall config file %s", err)
```
Comment (barnabepilliaudin) [inline back/config/config.go:53]: Shouldn't you use one of the unable you created in the constants you created, or create a new constant

File: back/config/config.go [hunk 4]
```diff
@@ -13,115 +14,53 @@ import (
...
+	}
 
-		if err = InitDatabase(config, productionEnv); err != nil {
-			return Config{}, errors.Wrapf(err, "buildEnvironmentConfig, unable to initialize production database")
-		}
+	config.Processor = configProcessor
 
-		log.Info("Production Environment has been loaded")
+	if err = initDatabase(config, env); err != nil {
+		return Config{}, errors.Wrapf(err, "buildEnvironmentConfig, unable to initialize database")
```
Comment (barnabepilliaudin) [inline back/config/config.go:59]: Shouldn't you use one of the unable you created in the constants you created, or create a new constant

File: back/config/config.go [hunk 5]
```diff
@@ -13,115 +14,53 @@ import (
...
-		err = InitDatabase(config, stagingEnv)
-		if err != nil {
-			return Config{}, errors.Wrapf(err, "buildEnvironmentConfig, unable to initialize staging database")
-		}
-
-		log.Info("Staging Environment has been loaded")
+func buildEnvironmentConfig(config Config, env string, configProcessor any) (Config, error) {
+	configFile, err := os.ReadFile(fmt.Sprintf("config/%s.yaml", env))
+	if err != nil {
+		return Config{}, errors.Errorf("loadConfig, unable to read config file %s", err)
```
Comment (barnabepilliaudin) [inline back/config/config.go:49]: Shouldn't you use one of the unable you created in the constants you created, or create a new constant

File: back/config/config.go [hunk 6]
```diff
@@ -13,115 +14,53 @@ import (
...
 	config.Handler, err = loadClientHandler(config)
 	if err != nil {
 		return Config{}, errors.Wrapf(err, "InitializeConfig, unable to load client handler")
 	}
 
-	config, err = buildEnvironmentConfig(config)
+	initializedConfig, err := selectConfigProcessor[config.Env.BuildProduction](config)
 	if err != nil {
-		return Config{}, errors.Wrapf(err, "InitializeConfig, unable to build environment config")
+		return Config{}, errors.Wrapf(err, "buildEnvironmentConfig, unable to build")
```
Comment (barnabepilliaudin) [inline back/config/config.go:40]: Shouldn't you use one of the unable you created in the constants you created, or create a new constant

File: back/config/config.go [hunk 7]
```diff
@@ -13,115 +14,53 @@ import (
...
-func InitializeConfig() (config Config, err error) {
 	envFile, err := os.ReadFile("config/default.yaml")
 	if err != nil {
-		return Config{}, errors.Wrapf(err, "InitializeConfig, unable to read config file")
+		return Config{}, errors.Errorf("InitializeConfig, unable to read config file %s", err)
 	}
 
 	if err = yaml.Unmarshal(envFile, &config); err != nil {
-		return Config{}, errors.Wrapf(err, "InitializeConfig, unable to unmarshal")
+		return Config{}, errors.Errorf("InitializeConfig, unable to unmarshall %s", err)
```
Comment (barnabepilliaudin) [inline back/config/config.go:26]: Shouldn't you use one of the unable you created in the constants you created, or create a new constant

File: back/config/config.go [hunk 8]
```diff
@@ -13,115 +14,53 @@ import (
...
-	SendgridToken string `yaml:"SENDGRID_TOKEN"`
-}
+func InitializeConfig() (Config, error) {
+	var config Config
 
-func InitializeConfig() (config Config, err error) {
 	envFile, err := os.ReadFile("config/default.yaml")
 	if err != nil {
-		return Config{}, errors.Wrapf(err, "InitializeConfig, unable to read config file")
+		return Config{}, errors.Errorf("InitializeConfig, unable to read config file %s", err)
```
Comment (barnabepilliaudin) [inline back/config/config.go:22]: Shouldn't you use one of the unable you created in the constants you created, or create a new constant

File: back/config/db.go [hunk 9]
```diff
@@ -68,24 +68,23 @@ func InitDatabase(config Config, env string) (err error) {
 
 		database, err = gorm.Open(postgres.Open(databaseURI))
 		if err != nil {
-			return errors.Errorf("InitDatabase, unable to load database %s", err)
+			return errors.Errorf("initDatabase, unable to load database %s", err)
```
Comment (barnabepilliaudin) [inline back/config/db.go:71]: Shouldn't you use one of the unable you created in the constants you created, or create a new constant

File: back/config/db.go [hunk 10]
```diff
@@ -68,24 +68,23 @@ func InitDatabase(config Config, env string) (err error) {
...
 		}
 
 	default:
 		database, err = InitNeonDatabase(config.Handler.VaultClient)
 		if err != nil {
-			return errors.Wrapf(err, "InitDatabase, unable to load database")
+			return errors.Wrapf(err, "initDatabase, unable to load database")
```
Comment (barnabepilliaudin) [inline back/config/db.go:77]: Shouldn't you use one of the unable you created in the constants you created, or create a new constant

File: back/config/db.go [hunk 11]
```diff
@@ -68,24 +68,23 @@ func InitDatabase(config Config, env string) (err error) {
...
 
 	log.Info("Database has been loaded")
 
-	if err := database.AutoMigrate(
+	if err := database.AutoMigrate(&model.Product{},
 		&model.Address{}, &model.Category{}, &model.Contact{},
 		&model.Material{}, &model.Order{}, &model.Product{},
 		&model.Status{}, &model.User{}); err != nil {
-		return errors.Errorf("InitDatabase, unable to auto migrate")
+		return errors.Errorf("initDatabase, unable to auto migrate")
```
Comment (barnabepilliaudin) [inline back/config/db.go:87]: Shouldn't you use one of the unable you created in the constants you created, or create a new constant

**Review by BarnabePILLIAUDIN — CHANGES_REQUESTED** [review]
That's look nice but you forgot to use the unable constant you create at some places

File: back/config/config.go (hunk 3 above)
Comment (cedricgautier) [reply back/config/config.go:53]: This is going to be used in a PR where I'll be integrating custom errors

File: back/config/config.go (hunk 4 above)
Comment (cedricgautier) [reply back/config/config.go:59]: This is going to be used in the next PR where I'll be integrating custom errors

File: back/config/db.go (hunk 11 above)
Comment (cedricgautier) [reply back/config/db.go:87]: This is going to be used in the next PR where I'll be integrating custom errors

File: back/config/db.go (hunk 10 above)
Comment (cedricgautier) [reply back/config/db.go:77]: This is going to be used in the next PR where I'll be integrating custom errors

File: back/config/db.go (hunk 9 above)
Comment (cedricgautier) [reply back/config/db.go:71]: This is going to be used in the next PR where I'll be integrating custom errors

File: back/config/config.go (hunk 8 above)
Comment (cedricgautier) [reply back/config/config.go:22]: This is going to be used in the next PR where I'll be integrating custom errors

File: back/config/config.go (hunk 7 above)
Comment (cedricgautier) [reply back/config/config.go:26]: This is going to be used in the next PR where I'll be integrating custom errors

File: back/config/config.go (hunk 6 above)
Comment (cedricgautier) [reply back/config/config.go:40]: This is going to be used in the next PR where I'll be integrating custom errors

File: back/config/config.go (hunk 5 above)
Comment (cedricgautier) [reply back/config/config.go:49]: This is going to be used in the next PR where I'll be integrating custom errors

File: back/config/config.go (hunk 4 above)
Comment (cedricgautier) [reply back/config/config.go:59]: This is going to be used in the next PR where I'll be integrating custom errors

**Review by BarnabePILLIAUDIN — APPROVED** [review]
Good!

**Review by avetisk — APPROVED** [review]

File: back/config/config.go (hunk 7 above)
Comment (KyDev25) [inline back/config/config.go:26]: Why did you change "errors.Wrapf" to "errors.Errorf"?

Comment (cedricgautier) [reply back/config/config.go:26]: It's switched to an ErrorF because it's not a parent function. 

You can check more about it  [here](https://github.com/KrispyTech/airneis/blob/staging/docs/coding-practices.md#error-handling)

Comment (KyDev25) [reply back/config/config.go:26]: Thank you for your documentation

**Review by KyDev25 — APPROVED** [review]
I really like the refactor, it will be very useful later on. The code is clear.
